src/hotspot/share/opto/callnode.cpp

```diff
@@ -1622,11 +1622,15 @@ Node *AllocateArrayNode::make_ideal_length(const TypeOopPtr* oop_type, PhaseTran
              "narrow type must be narrower than length type");
 
       // Return NULL if new nodes are not allowed
-      if (!allow_new_nodes) return NULL;
+      if (!allow_new_nodes) {
+        return NULL;
+      }
       // Create a cast which is control dependent on the initialization to
       // propagate the fact that the array length must be positive.
       InitializeNode* init = initialization();
-      assert(init != NULL, "initialization not found");
+      if (init == NULL) {
+        return NULL; // Return NULL if dead path
+      }
```

Comment:
Can all callers deal with NULL?  It looks like GraphKit::array_ideal_length() will crash if make_ideal_length() returns NULL.

This fix looks different from what you proposed in the bug report:
> Or attach CastII(length) to Allocate node control projection if Initialize node is not present. As I said before we may have legal cases like that (as I remember when working on EA).